Adds WorkflowJobTemplate Add and Edit form and functionality.

AlexSCorey opened the pull request:
##### SUMMARY
This addresses #5813 and https://github.com/ansible/awx/issues/5930

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
```
9.1.1
```


##### ADDITIONAL INFORMATION
![Screen Shot 2020-02-06 at 4 42 03 PM](https://user-images.githubusercontent.com/39280967/73981174-b3e1c700-48ff-11ea-97b4-c62ed6a9b79f.png)
![Screen Shot 2020-02-06 at 4 42 16 PM](https://user-images.githubusercontent.com/39280967/73981175-b47a5d80-48ff-11ea-9280-9bdadc81c075.png)

AlexSCorey committed:
Adds WorkflowJobTemplate Add and Edit form and functionality.

AlexSCorey committed:
Adds WorkflowJobTemplate Add and Edit form and functionality.

Fields work and forms save

Adds Webhook form fields and tooltip to VariablesField component

Alpha order

Fixes contentLoading issue in PR and enables Launch on JT List

Adds page section to fix render issue

Adds subform restore functionality and addresses PR issues

Adds pageSection to jt add form and fixes other PR issues

-Fixes spelling error on WFJTDetail
-Adds page section to JT Add Form to fix styling issue
-Adds spacing between functions
-Fixes form submission error by allowing state to handle the lookups while formik
only handles their ids.

Fixes styling issues, navigation, props, and adds useRequest hook
-Add functionality to remove chips from look up fields
-Removes uncessary custome styling from
-Removes uncessary Form Group wrappers
-Adds internationalization to webhook key string.
-Adds field level error handling
-updates tests
-Adds initial null value to form submit error

Adds error handling to submit labels, prevents uncessary api call

The unecessary api call is for the webhook credential id.  If there is no
webhook service we do not want the api to make a call for get the webhook credential
type id.

Adds error handling test to add and edit form. Updates Form component

Updates tests to a real error.

Fixes extra vars bug

AlexSCorey commented on awx/ui_next/src/api/models/WorkflowJobTemplates.js:
```diff
@@ -7,7 +7,23 @@ class WorkflowJobTemplates extends Base {
 
   readNodes(id, params) {
-    return this.http.get(`${this.baseUrl}${id}/workflow_nodes/`, { params });
+    return this.http.get(`${this.baseUrl}${id}/workflow_nodes/`, {
```
I will put these functions in alpha order

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateAdd/WorkflowJobTemplateAdd.jsx:
```diff
@@ -0,0 +1,49 @@
+import { WorkflowJobTemplatesAPI } from '@api';
+import WorkflowJobTemplateForm from '../shared/WorkflowJobTemplateForm';
+
+function WorkflowJobTemplateAdd() {
```
Will add spacing between functions so it isn't so dense.

AlexSCorey commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateEdit/WorkflowJobTemplateEdit.jsx:
```diff
@@ -0,0 +1,76 @@
+      template.summary_fields.labels.results,
+      labels
+    );
+    if (!orgId && !template.organization) {
```
spoke to @jlmitch5  about this use case, which is if no Org is selected in the form and there is no org associated with the template and I generate a label (which needs an org ID to post to api) which org do I use.  Looks like old ui makes a call to the api, gets list of orgs and picks the first one in the list. Seems weird to me so I definitely open to suggestions about this.

jlmitch5 replied:
I think org is just a required field on the label model, even though for all intents and purposes, it doesn't seem necessary.

AlexSCorey commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,176 @@
+        variables: template.variables || '---',
+        limit: template.limit || '',
+        scmBranch: template.scm_branch || '',
+        ask_limit_on_launch: template.ask_limit_on_launch || false,
```
We actually don't need these yet, prompt on launch is not functional yet, so I'll remove them.

mabashian replied:
@AlexSCorey I think these are fine to leave in.  Shouldn't be any negative impact that I can see.

marshmalien reviewed (changes requested):
Pulled this down and poked around - saving a new workflow and editing an existing workflow works great! 

We are missing a few form items: 
* SCM Branch field should have a (?) helper tooltip
* Options checkboxes: Enable Concurrent Jobs and Enable Webhook
* ~Launch button~ (Scratch that - we only need Launch in Details 😅)
* Surface form errors with `FormSubmitError` (example in HostForm)

marshmalien commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,176 @@
+                  helperTextInvalid={form.errors.inventory}
+                  onChange={value => {
+                    form.setFieldValue('inventory', value?.id || null);
+                    form.setFieldValue(
```
Do we need to update `organizationId` in the inventory lookup? What does `organizationId` do? We need an organization id to save labels on a workflow template?

AlexSCorey replied:
when we generate a label in the form the api requires an org id to get be pass along with the label name.

marshmalien commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.test.jsx:
```diff
@@ -0,0 +1,137 @@
+      await wrapper.find('button[aria-label="Save"]').simulate('click');
+    });
+
+    act(() => {
```
We don't need to wrap this assertion in `act()`

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplate.jsx:
```diff
@@ -94,58 +95,72 @@ class WorkflowJobTemplate extends Component {
+                path="/templates/workflow_job_template/:id/edit"
+                render={() => (
+                  <WorkflowJobTemplateEdit
+                    hasTemplateLoading={hasContentLoading}
```
We are using the prop `hasContentLoading` in Workflow Edit, not `hasTemplateLoading`. Lets update the loading prop names to match.

mabashian commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,176 @@
+                <InventoryLookup
+                  value={inventory}
+                  onBlur={() => form.setFieldTouched('inventory')}
+                  tooltip={i18n._(t`Select the inventory containing the hosts
```
We probably want this tooltip to say something about the fact that the inventory will override the inventory for all nodes.

<img width="480" alt="Screen Shot 2020-02-12 at 10 41 16 AM" src="https://user-images.githubusercontent.com/9889020/74350820-4c56cc00-4d84-11ea-8065-54a1139a0c3a.png">

When in doubt, I'd match up the tooltips for form fields with whatever we have in the existing app.

AlexSCorey replied:
👍

AlexSCorey committed:
Fields work and forms save

AlexSCorey committed:
Adds Webhook form fields and tooltip to VariablesField component

AlexSCorey commented:
Recheck

AlexSCorey commented:
recheck

jlmitch5 commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,454 @@
+  submitError: null,
+};
+
+export default withI18n()(WorkflowJobTemplateForm);
```
nice, glad this is functional, one less thing we'll have to convert!

jlmitch5 reviewed (commented):
looking great!

AlexSCorey committed:
Alpha order

AlexSCorey committed:
Fixes contentLoading issue in PR and enables Launch on JT List

marshmalien commented on awx/ui_next/src/screens/Template/Templates.jsx:
```diff
@@ -53,47 +56,49 @@ class Templates extends Component {
     return (
       <>
         <Breadcrumbs breadcrumbConfig={breadcrumbConfig} />
-        <PageSection>
```
I think we either need to keep PageSection component at the top level, or wrap all of the children components in PageSection.

AlexSCorey replied:
I think we need to add PageSection as close to the actual component that renders the view as possible.  I'm in favor of wrapping the children as I think it'll be easier to reuse views, thinking nested views, in the future.

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplate.jsx:
```diff
@@ -47,19 +49,21 @@ class WorkflowJobTemplate extends Component {
         data.summary_fields.webhook_credential.kind = name;
       }
-
+      setBreadcrumb(data);
```
Line 66 and 68 are duplicates.

marshmalien commented on awx/ui_next/src/screens/Template/WorkflowJobTemplate.jsx:
```diff
@@ -116,58 +122,72 @@ class WorkflowJobTemplate extends Component {
-                />
-              )}
+      <PageSection>
+        <Card className="awx-c-card">
```
Remove this className - Those classes have been deleted.

AlexSCorey committed:
Adds page section to fix render issue

marshmalien commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,454 @@
+            <VariablesField
+              id="wfjt-variables"
+              name="variables"
+              label={i18n._(t`Variables`)}
```
Should we pass a tooltip here?

mabashian replied:
I don't think this field takes a tooltip yet

mabashian replied:
But definitely should in the future - I'll create an issue @marshmalien

marshmalien replied:
We're passing a tooltip prop to VariablesField here in this PR: https://github.com/ansible/awx/pull/5841/files#diff-ec5b74cb8508843457f55569b7a8eef9R25

It looks like the Inventory Form is using this prop to show a tooltip on the VariablesField.

AlexSCorey replied:
@mabashian I can just add the tooltip from the old ui in there, unless the tooltip needs to be updated or something?  I intended to do that anyway and just skipped over it.

mabashian commented on awx/ui_next/src/screens/Template/WorkflowJobTemplate.jsx:
```diff
@@ -47,19 +49,21 @@ class WorkflowJobTemplate extends Component {
           data: { webhook_key },
         } = await WorkflowJobTemplatesAPI.readWebhookKey(id);
-        this.setState({ webHookKey: webhook_key });
+        this.setState({ webhook_key });
```
I'm not sure if this matters or not - but I expected this variable to be initialized up in the constructor with all the other state variables.

AlexSCorey replied:
I don't know that it matters either, but I can add it with a null value.

mabashian commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,454 @@
+            label={i18n._(t`Options`)}
+            css="margin-top: 20px"
+          >
+            <Checkbox
```
I'd recommend taking a look at the `CheckboxField` component.  I think it shorten up this options code a bit.  We started using it in the job template form for the options section. https://github.com/ansible/awx/blob/devel/awx/ui_next/src/screens/Template/shared/JobTemplateForm.jsx#L525

AlexSCorey committed:
Adds subform restore functionality and addresses PR issues

AlexSCorey commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,473 @@
+              label={i18n._(t`Webhooks`)}
+            >
+              {({ form }) => (
+                <Checkbox
```
I was unable to use the CheckboxField component here because I needed access the the onChange function.  This field manages whether the subform is open, and the state of the subform when the user opens and closes the form.

AlexSCorey commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,473 @@
+  const setWebhookValues = (form, webhookServiceValue) => {
+    if (
+      webhookServiceValue === form.initialValues.webhook_service ||
+      webhookServiceValue === ''
```
this || applies to the use case where the user hasWebhooks, (on editing a form), they change the service, and then disabled the webhooks, then re-enable webhooks, we want it to reset to the initial value.

AlexSCorey commented on awx/ui_next/src/screens/Template/shared/WorkflowJobTemplateForm.jsx:
```diff
@@ -0,0 +1,473 @@
+    if (!enabledWebhooks) {
+      initialWebhookKey = webhookKey;
+      form.setFieldValue('webhook_credential', null);
+      form.setFieldValue('webhook_service', '');
```
I think for https://github.com/ansible/awx/pull/5841/files#r383352280 I can actually reset this to the initial value and will be able to remove the || here https://github.com/ansible/awx/pull/5841/files#r383352280

AlexSCorey committed:
Adds pageSection to jt add form and fixes other PR issues

-Fixes spelling error on WFJTDetail
-Adds page section to JT Add Form to fix styling issue
-Adds spacing between functions
-Fixes form submission error by allowing state to handle the lookups while formik
only handles their ids.

AlexSCorey committed:
Fixes styling issues, navigation, props, and adds useRequest hook
-Add functionality to remove chips from look up fields
-Removes uncessary custome styling from
-Removes uncessary Form Group wrappers
-Adds internationalization to webhook key string.
-Adds field level error handling
-updates tests
-Adds initial null value to form submit error

AlexSCorey committed:
Adds error handling to submit labels, prevents uncessary api call

The unecessary api call is for the webhook credential id.  If there is no
webhook service we do not want the api to make a call for get the webhook credential
type id.

AlexSCorey commented:
recheck

AlexSCorey committed:
Adds error handling test to add and edit form. Updates Form component

AlexSCorey committed:
Updates tests to a real error.

marshmalien reviewed (approved):
🎉

mabashian committed:
Merge branch 'devel' into 5813-WorkflowJTForm

dsesami commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateDetail/WorkflowJobTemplateDetail.jsx:
```diff
@@ -55,7 +58,7 @@ function WorkflowJobTemplateDetail({ template, i18n, webHookKey }) {
       {template.webhook_service && (
         <TextListItem component={TextListItemVariants.li}>
-          {i18n._(t`- Webhooks`)}
+          {i18n._(t`- Enable Webhook`)}
```
Enable Webhook (and its respective tooltip) should both say "Enable Webhooks". Might as well change the aria-label etc. as well if you do this.

mabashian replied:
Yea this was done intentionally to match the old UI

<img width="1508" alt="Screen Shot 2020-03-04 at 9 47 17 AM" src="https://user-images.githubusercontent.com/9889020/75891019-2f4e6f80-5dfd-11ea-9498-32239fccb218.png">

I would defer to @wenottingham / @tvo318 on this one

AlexSCorey replied:
@mabashian and @dsesami Based on their feedback let's make a follow up issue....might be good to mark it as good first issue also.

tvo318 replied:
IMO, I agree with @dsesami - should be plural, since the old one was plural: “Enable Webhooks”

dsesami replied:
https://github.com/ansible/awx/issues/6163

dsesami commented on awx/ui_next/src/components/CodeMirrorInput/VariablesField.jsx:
```diff
@@ -20,7 +20,15 @@ const StyledCheckboxField = styled(CheckboxField)`
 
-function VariablesField({ i18n, id, name, label, readOnly, promptId }) {
+function VariablesField({
+  i18n,
```
On a new WF Template Form:
1. Go to variables
2. Set some vars in the YAML field, like `foo: bar`. Save.
3. Go to edit the field again.

The current value vanishes, as opposed to having kept it for editing.

mabashian replied:
This sounds like a legitimate bug.  If this is unique to WFJT's then lets fix it in this branch.  If it impacts other extra vars in other forms then lets file a bug and move forward.

AlexSCorey replied:
yeah that was a legitimate bug and has been resolved.

dsesami replied:
@AlexSCorey this bug was within the latest branch I pulled this morning. Is it resolved in another branch? If so, could you rebase so that the bug is fixed here? thanks!

dsesami replied:
I've confirmed that this only happens with WFJTs as well.

dsesami replied:
This has been fixed, thanks.

dsesami commented on awx/ui_next/src/screens/Template/WorkflowJobTemplateDetail/WorkflowJobTemplateDetail.jsx:
```diff
@@ -55,7 +58,7 @@ function WorkflowJobTemplateDetail({ template, i18n, webHookKey }) {
       {template.webhook_service && (
         <TextListItem component={TextListItemVariants.li}>
-          {i18n._(t`- Webhooks`)}
+          {i18n._(t`- Enable Webhook`)}
```
On the add WF template form, when you are creating a Webhook configuration, the Webhook URL and Webhook Key fields are missing. These should be visible, but have the pending status, like so:

![Screenshot from 2020-03-04 09-40-31](https://user-images.githubusercontent.com/19942822/75890324-345eef00-5dfc-11ea-9a97-4a8b8d4a48f7.png)

mabashian replied:
Yea in talking with @AlexSCorey I think this was done intentionally on the add form since those fields are generated by the backend.  In limited testing of the old UI I think it behaves the same way(?).  If so, lets file a follow-up enhancement for the UX pattern here.

AlexSCorey replied:
@dsesami one of the problems with have the url field visible in this state is that the id of the template is required to generate an accurate url.  Without the id the url will be something like `/api/v2/workflow_job_template/undefined/github` and I was of the opinion that was not good form, and it doesn't follow the pattern in the old ui.  In the old ui both of those fields are not displayed on the add form.  

In either case, if this is necessary I agree with Mike, let's make a follow up issue.

dsesami replied:
Yeah, I made a similar suggestion with the old UI. I'll create a follow-up ticket.

dsesami replied:
ticket here: https://github.com/ansible/awx/issues/6162

dsesami reviewed (changes requested):
I've left a few comments, some blocking, some not. One other suggestion: within the scope of this PR, is it possible to include tooltips on the Details page itself, and not only when editing the form? If it is not in scope, let me know so I can make an issue about it.

AlexSCorey commented:
> I've left a few comments, some blocking, some not. One other suggestion: within the scope of this PR, is it possible to include tooltips on the Details page itself, and not only when editing the form? If it is not in scope, let me know so I can make an issue about it.

We have not had tooltips in the details view in the rewrite.  It's probably a good idea, but would not be within the scope of this PR.

dsesami reviewed (approved):
Looks great. I've created follow-up issues on all agreed-upon items.
